src/Dom/Document.php

```diff
@@ -2145,4 +2147,22 @@ private function addHttpEquivCharset($html)
 
         return $html;
     }
+
+    /**
+     * Check if there are malformed byte sequences.
+     *
+     * If passed bad byte sequences DOMDocument would fail in a random way (silently) and produces garbage.
+     *
+     * @param string $source The HTML fragment string.
+     * @throws \InvalidArgumentException if source contains some malformed byte sequences
+     */
+    private function assertNoneMalformedByteSequences($source)
+    {
+        if ($this->options[Option::ADDITIONAL_VALIDATION] && ! mb_check_encoding($source)) {
+            throw new \InvalidArgumentException(
```

Comment:
This should use a custom exception `InvalidByteSequence` with a named constructor (see `src/Exception`).

Reply:
Should the exception message contain an HTML source?
If not, is there necessary to write a test for InvalidByteSequence that contains only a message?

Reply:
Normally yes, but for string-based HTML, it is actually hard to produce something useful for an exception message that doesn't flood the screen/logs... Not sure how we would go about that here.

Reply:
The third way is to add a second argument to the named constructor and then provide a method like `$invalidByteSequence->getHtmlSource()` to get the html contained invalid byte sequence.